pkg/metrics.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright DDEV Technologies LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pkg
+
+import (
+	"k8s.io/klog"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	segment "gopkg.in/segmentio/analytics-go.v3"
+)
+
+// Prometheus metrics for bot command invocations
+var (
+	CreatePreviewSiteCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "bot_create_preview_site_total",
+		Help: "Total number of create preview site command invocations",
+	}, []string{"bot", "status"})
+
+	DeletePreviewSiteCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "bot_delete_preview_site_total",
+		Help: "Total number of delete preview site command invocations",
+	}, []string{"bot", "action", "status"})
+
+	HelpPreviewSiteCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "bot_help_preview_site_total",
+		Help: "Total number of help command invocations",
+	}, []string{"bot", "action"})
+
+	// Label for counters that can be triggered either by command directly or action in a repository
+	metricLabelCommand = "command"
+
+	// Labels for site creation status
+	metricLabelPreviewSuccess            = "success"
+	metricLabelPreviewFailed             = "internal error"
+	metricLabelPreviewDeniedMissingEmail = "failed to verify user email"
+	metricLabelPreviewAuthAPIError       = "failed to query admin API"
+	metricLabelPreviewDenied             = "unauthorized"
+	metricLabelPreviewNothingToDelete    = "no preview site to be deleted"
+)
+
+type analytics struct {
+	botType string
+	appId   string
+	cli     segment.Client
+}
+
+type noopLog struct{}
+
+func (n *noopLog) Logf(f string, args ...interface{})   {}
+func (n *noopLog) Errorf(f string, args ...interface{}) {}
+
+func AnalyticsClient(botType, appId, writeKey string) *analytics {
+	if writeKey == "" {
+		klog.Info("analytics client not configured")
+		return &analytics{
+			botType: botType,
+			appId:   appId,
+		}
+	}
+	cli, err := segment.NewWithConfig(writeKey, segment.Config{Logger: &noopLog{}})
+	if err != nil {
+		klog.Errorf("failed to init analytics client: %v", err)
+	}
+	return &analytics{
+		botType: botType,
+		appId:   appId,
+		cli:     cli,
+	}
+}
+
+func (a *analytics) CreatePreviewSite(user, result string) {
```

Comment:
_Perhaps just thinking out loud here._

Not a huge fan of these function names but struggle to think of one that resolves my conflict with them.  They are named in such a way to perform an action whereas they only record metrics, suppose that it could be gathered from the structure in which the functions belong.  `IncCreatePreview`  `ObserveCreatePreview`  Or could have seen a singular `Observe` function which could take a:
```
func (a *analytics) Observe(cmd BotCommand, user string, result string){
```
if we typed them:
```
type BotCommand string
const PreviewSite BotCommand = "/ddev-live-preview-site"
```
https://github.com/drud/repo-chat-bot/blob/main/pkg/responses.go#L46

However from my understanding PreviewSite is just a string and we don't really need to type them also.

Reply:
I like the `Observe` prefix and maybe, even more, the singular `Observer()` function, will see how these two look and pick one that I like better.

Reply:
so because these functions take slightly different arguments, I decided to go with the `Observe` prefix.